### PR TITLE
JDK-8319314: NMT detail report slow or hangs for large number of mappings

### PR DESCRIPTION
**Pull request by tstuefe**
If a lot of mappings have been registered with NMT, NMT detail report may get so slow as to hang seemingly. 

100 k mappings will take > 1 minute
1 million mappings never returns (I canceled the test after 90 minutes)

We burn all the time doing a sorted insert into the target array of the MemBaseline object. That sorted insert is pointless since the source array is already sorted by region base address.

Omitting the pointless sort drops the runtime for 1 million mappings from > 90 minutes to 48 seconds.

Tests: Tested locally. GHAs (Windows error unrelated).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319314](https://bugs.openjdk.org/browse/JDK-8319314): NMT detail report slow or hangs for large number of mappings (**Bug** - P4)

### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**) ⚠️ Review applies to [b1ad3728](https://git.openjdk.org/jdk/pull/16480/files/b1ad37286c2e8a2987e7aa92993f84dfd5eda469)
 * [Gerard Ziemski](https://openjdk.org/census#gziemski) (@gerard-ziemski - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16480/head:pull/16480` \
`$ git checkout pull/16480`

Update a local copy of the PR: \
`$ git checkout pull/16480` \
`$ git pull https://git.openjdk.org/jdk.git pull/16480/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16480`

View PR using the GUI difftool: \
`$ git pr show -t 16480`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16480.diff">https://git.openjdk.org/jdk/pull/16480.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16480#issuecomment-1792361078)

### CODE REVIEW

File: src/hotspot/share/nmt/memBaseline.cpp [hunk 1]
```diff
@@ -110,22 +110,24 @@ class MallocAllocationSiteWalker : public MallocSiteWalker {
...
-  VirtualMemoryAllocationWalker() : _count(0) { }
+  VirtualMemoryAllocationWalker() :
+    _count(0)
+#ifdef ASSERT
+    , _last_base(nullptr)
+#endif
+  {}
 
   bool do_allocation_site(const ReservedMemoryRegion* rgn)  {
+    assert(rgn->base() >= _last_base, "region unordered?");
```
Comment (jdksjolen) [inline src/hotspot/share/nmt/memBaseline.cpp:130]: It seems like we don't assign to `_last_base`?

**Review by jdksjolen — APPROVED** [review]
Hi,

Thanks for this, one small thing to be fixed.

**Comment by tstuefe** [comment]
Ping @gerard-ziemski ?

**Comment by tstuefe** [comment]
Ping anyone?

**Comment by gerard-ziemski** [comment]
Taking a look...

**Review by gerard-ziemski — CHANGES_REQUESTED** [review]
I have a couple small questions/feedback.

Also, you said:

```
100 k mappings will take > 1 minute
1 million mappings never returns (I canceled the test after 90 minutes)
```

What are those times now with this change?

**Comment by tstuefe** [comment]
Hi Gerard,

thanks for looking into this. Answers below.

> I have a couple small questions/feedback - see below.
> 
> Also, you said:
> 
> ```
> 100 k mappings will take > 1 minute
> 1 million mappings never returns (I canceled the test after 90 minutes)
> ```
> 
> What are those times now with this change?

1 million mappings finished with my patch in 48 seconds.
100k mappings finished in 19 seconds I believe? The growth was exponential before my patch, now it seems linear.

File: src/hotspot/share/nmt/memBaseline.cpp [hunk 2]
```diff
@@ -110,22 +110,25 @@ class MallocAllocationSiteWalker : public MallocSiteWalker {
...
+  EntryList _virtual_memory_regions;
+  size_t    _count;
+  DEBUG_ONLY(address _last_base;)
  public:
-  VirtualMemoryAllocationWalker() : _count(0) { }
+  VirtualMemoryAllocationWalker() :
+    _count(0)
+#ifdef ASSERT
+    , _last_base(nullptr)
+#endif
```
[the remark this answers is not in the document]
Comment (tstuefe) [reply src/hotspot/share/nmt/memBaseline.cpp:126]: I think that won't compile. I don't remember the exact reason, but DEBUG_ONLY does not work for initializer lists.

File: src/hotspot/share/nmt/memBaseline.cpp [hunk 3]
```diff
@@ -110,22 +110,25 @@ class MallocAllocationSiteWalker : public MallocSiteWalker {
...
 class VirtualMemoryAllocationWalker : public VirtualMemoryWalker {
  private:
-  SortedLinkedList<ReservedMemoryRegion, compare_virtual_memory_base>
-                _virtual_memory_regions;
-  size_t        _count;
-
+  typedef LinkedListImpl<ReservedMemoryRegion, AnyObj::C_HEAP, mtNMT,
+                         AllocFailStrategy::RETURN_NULL> EntryList;
+  EntryList _virtual_memory_regions;
+  size_t    _count;
```
[the remark this answers is not in the document]
Comment (tstuefe) [reply src/hotspot/share/nmt/memBaseline.cpp:119]: No idea. I guess this is a remnant, and could probably be removed, but ideally in another cleanup PR.

**Comment by tstuefe** [comment]
@gerard-ziemski Do you want me to make any other changes? About the _count, I don't really want to touch that since I would like to keep this change small minimal and focused, since it needs to be backported to at least 21 and 17, possibly further down.

**Review by gerard-ziemski — APPROVED** [review]

**Comment by tstuefe** [comment]
Thanks @gerard-ziemski and @jdksjolen !

/integrate

**Pull request closed without merging**